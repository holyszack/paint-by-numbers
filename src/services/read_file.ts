import { Observable, Subscriber } from "rxjs";

export const readFile = (file: File | Blob): Observable<Buffer> => Observable.create((obs: Subscriber<Buffer>) => {
    if (!(file instanceof File) && !(file instanceof Blob)) {
        obs.error(new Error('`blob` must be an instance of File or Blob.'));
        return;
    }

    const reader = new FileReader();

    reader.onerror = (err) => obs.error(err);
    reader.onabort = (err) => obs.error(err);
    reader.onload = () => reader.result instanceof ArrayBuffer
        ? obs.next(Buffer.from(reader.result))
        : obs.error("No Data Found");
    reader.onloadend = () => obs.complete();

    return reader.readAsArrayBuffer(file);
});
