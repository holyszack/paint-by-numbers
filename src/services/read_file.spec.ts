import { readFile } from "./read_file";
import fs from "fs";

// tslint:disable:max-line-length
const tests = [
    {
        "expected": "",
        "value": new Blob([Uint8Array.from(fs.readFileSync("./coverage/test.png"))]),
    },
];
// tslint:enable:max-line-length
describe("readFile [unit]", () => {
    tests.forEach(({ expected, value }) => {
        it(`${value} should equal ${expected}`, (done) => {
            const results: Buffer[] = [];
            readFile(value).subscribe(
                (a) => results.push(a),
                (e) => { console.error(e); done(); },
                () => {
                    expect(results).toMatchSnapshot();
                    done();
                },
            );
        });
    });
});
