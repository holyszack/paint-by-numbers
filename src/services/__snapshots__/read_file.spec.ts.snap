// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readFile [unit] [object Blob] should equal  1`] = `
Array [
  Object {
    "data": Array [
      137,
      80,
      78,
      71,
      13,
      10,
      26,
      10,
      0,
      0,
      0,
      13,
      73,
      72,
      68,
      82,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      1,
      8,
      2,
      0,
      0,
      0,
      144,
      119,
      83,
      222,
      0,
      0,
      0,
      9,
      112,
      72,
      89,
      115,
      0,
      0,
      46,
      35,
      0,
      0,
      46,
      35,
      1,
      120,
      165,
      63,
      118,
      0,
      0,
      0,
      7,
      116,
      73,
      77,
      69,
      7,
      227,
      4,
      14,
      19,
      5,
      37,
      97,
      72,
      47,
      84,
      0,
      0,
      0,
      25,
      116,
      69,
      88,
      116,
      67,
      111,
      109,
      109,
      101,
      110,
      116,
      0,
      67,
      114,
      101,
      97,
      116,
      101,
      100,
      32,
      119,
      105,
      116,
      104,
      32,
      71,
      73,
      77,
      80,
      87,
      129,
      14,
      23,
      0,
      0,
      0,
      12,
      73,
      68,
      65,
      84,
      8,
      215,
      99,
      248,
      223,
      192,
      0,
      0,
      4,
      1,
      1,
      128,
      20,
      9,
      7,
      2,
      0,
      0,
      0,
      0,
      73,
      69,
      78,
      68,
      174,
      66,
      96,
      130,
    ],
    "type": "Buffer",
  },
]
`;
